File: src/main/java/org/cip4/xjdf/json/openapi/SimpleTypeLocalList.java
package org.cip4.xjdf.json.openapi;

import org.cip4.xjdf.json.openapi.model.Model;
import org.cip4.xjdf.json.openapi.model.Schema;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;

public class SimpleTypeLocalList extends SimpleType {

    private final Context context;

    public SimpleTypeLocalList(Node node, Context context) {
        super(node);
        this.context = context;
    }

    private AttributeType getBase() {
        try {
            Node baseNode = (Node) context.getXPath().evaluate(
                "xs:list",
                node,
                XPathConstants.NODE
            );
            return AttributeType.create(baseNode, context);
        } catch (Exception e) {
            throw new RuntimeException("Error evaluating base node", e);
        }
    }

    @Override
    public Model getModel() {
        return new Model(
            getName(),
            new Schema().type("array").items(getBase().getModel().getSchema())
        );
    }

    public static class Factory implements SimpleType.Factory<SimpleTypeLocalList> {
        @Override
        public boolean supports(Node candidate, XPath xPath) {
            try {
                NodeList nodes = (NodeList) xPath.evaluate(
                    "xs:list/xs:simpleType",
                    candidate,
                    XPathConstants.NODESET
                );
                return nodes.getLength() > 0;
            } catch (Exception e) {
                throw new RuntimeException("Error evaluating supports condition", e);
            }
        }

        @Override
        public SimpleTypeLocalList create(Node node, Context context) {
            return new SimpleTypeLocalList(node, context);
        }
    }
}

File: src/main/kotlin/org/cip4/xjdf/openapi/SimpleTypeLocalList.kt
package org.cip4.xjdf.openapi

import org.cip4.xjdf.openapi.model.Schema
import org.w3c.dom.Node
import org.w3c.dom.NodeList
import javax.xml.xpath.XPath
import javax.xml.xpath.XPathConstants

class SimpleTypeLocalList(
    node: Node,
    private val context: Context
) : SimpleType(node) {

    private val base: AttributeType
        get() = AttributeType.Factory.create(
            context.xPath.evaluate(
                "xs:list",
                node,
                XPathConstants.NODE
            ) as Node,
            context
        )

    companion object Factory : SimpleType.Factory<SimpleTypeLocalList> {
        override fun supports(candidate: Node, xPath: XPath): Boolean =
            (xPath.evaluate(
                "xs:list/xs:simpleType",
                candidate,
                XPathConstants.NODESET
            ) as NodeList).length > 0

        override fun create(node: Node, context: Context): SimpleTypeLocalList =
            SimpleTypeLocalList(node, context)
    }

    override fun getModel(): Schema {
        return Schema(type = "array", items = base.getModel())
    }
}